lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -2025,10 +2028,14 @@ Attribute DotOperandEncodingAttr::parse(AsmParser &parser, Type type) {
 }
 
 void DotOperandEncodingAttr::print(mlir::AsmPrinter &printer) const {
-  auto mmaParent = getParent().dyn_cast<NvidiaMmaEncodingAttr>();
+  auto parent = getParent();
+  auto mmaParent = parent.dyn_cast<NvidiaMmaEncodingAttr>();
   printer << "<{"
           << "opIdx = " << getOpIdx() << ", parent = " << getParent();
-  if (mmaParent && mmaParent.isAmpere())
```

Comment:
I have a question related to Nvidia.

This part mismatches parsing:
- Parse for Ampere and Hopper
- Print only for Ampere

Can anyone clarify, what behavior is correct?

Reply:
I'm 60% sure that printing for Hopper is correct, but also in the 40% case that it's incorrect then printing it will make it obvious to someone else that it shouldn't be there.  :)  Thanks for noticing this.

Reply:
Also this whole thing seems unnecessarily confusing.  It seems to me that we shouldn't need any of this custom parsing logic.  Instead:

- Make kWidth an optional attribute with a default of 0 or -1.
- Use a regular MLIR attr parser and printer, no need for the custom stuff.  It will not print kWidth when it's 0, because that's the default.
- Add a verifier that checks that kWidth is 0 for those encodings that don't support kWidth.  Also check that the kWidth is explicitly equal to `AMDWmmaEncodingAttr::getMNKDimPerWMMAInstr()[2]` for AMDWmma.

Not your sin, I guess you don't need to change it.

Reply:
It is OK to change formatting.  You'll have to fix the tests, but I've done it, it's not too bad.

Reply:
Turned out declarative format can express same formatting :slightly_smiling_face:

Reply:
Yup. Great to see this getting cleaned up. Thanks @binarman!